File: jesd204b/trunk/software/jesd204b/JesdTx.cpp
//-----------------------------------------------------------------------------
// File          : JesdTx.cpp
// Author        : Uros legat <ulegat@slac.stanford.edu>
//                            <uros.legat@cosylab.com>
// Created       : 27/04/2015
// Project       : 
//-----------------------------------------------------------------------------
// Description :
//    Device container for Jesd204b
//-----------------------------------------------------------------------------
// Copyright (c) 2015 by SLAC. All rights reserved.
// Proprietary and confidential to SLAC.
//-----------------------------------------------------------------------------
// Modification history :
// 27/04/2015: created
//-----------------------------------------------------------------------------
#include <JesdTx.h>
#include <Register.h>
#include <RegisterLink.h>
#include <Variable.h>
#include <Command.h>
#include <sstream>
#include <iostream>
#include <string>
#include <iomanip>
using namespace std;

// Constructor
JesdTx::JesdTx ( uint32_t linkConfig, uint32_t baseAddress, uint32_t index, Device *parent, uint32_t addrSize ) : 
                        Device(linkConfig,baseAddress,"JesdTx",index,parent) {

   // Description
   desc_ = "Common JESD interface object.";

   // Create Registers: name, address
   RegisterLink *rl;
   
   addRegisterLink(rl = new RegisterLink("Enable",           baseAddress_ + (0x00*addrSize), Variable::Configuration));
   rl->getVariable()->setDescription("Enables the Tx modules: 0x3 - enables both modules at a time");
   
   
   // addRegisterLink(rl = new RegisterLink("SysrefDelay",      baseAddress_ + (0x01*addrSize), Variable::Configuration));
   // rl->getVariable()->setDescription("Sets the synchronisation delay in clock cycles");

   // addRegisterLink(rl = new RegisterLink("AXISTrigger",      baseAddress_ + (0x02*addrSize), Variable::Configuration));
   // rl->getVariable()->setDescription("Triggers the AXI stream transfer: 0x3 - triggers both modules at a time");
   
   
   // addRegisterLink(rl = new RegisterLink("AXISpacketSize",   baseAddress_ + (0x03*addrSize), Variable::Configuration));
   // rl->getVariable()->setDescription("Data packet size (when enabled packets are being sent continuously)"); 

   addRegisterLink(rl = new RegisterLink("CommonControl",    baseAddress_ + (0x04*addrSize), 1, 5,
                                "SubClass",              Variable::Configuration, 0, 0x1,
                                "ReplaceEnable",         Variable::Configuration, 1, 0x1,
                                "ResetGTs",              Variable::Configuration, 2, 0x1,
                                //"ClearErrors",           Variable::Configuration, 3, 0x1,
                                "InvertSync",            Variable::Configuration, 4, 0x1,
                                "EnableTestSig",         Variable::Configuration, 5, 0x1));

   addRegisterLink(rl = new RegisterLink("PeriodRampStep",    baseAddress_ + (0x05*addrSize), 1, 2,
                                "RampStep",              Variable::Configuration, 0, 0xffff,
                                "SquarePeriod",          Variable::Configuration, 16,0xffff));
   rl->getVariable()->setDescription("Ramp increment step and a period of the wave in c-c ");

   addRegisterLink(rl = new RegisterLink("LowAmplitudeVal",   baseAddress_ + (0x06*addrSize), Variable::Configuration));
   rl->getVariable()->setDescription("Low value of the square waveform amplitude");

   addRegisterLink(rl = new RegisterLink("HighAmplitudeVal",  baseAddress_ + (0x07*addrSize), Variable::Configuration));
   rl->getVariable()->setDescription("High value of the square waveform amplitude");
   
   addRegisterLink(rl = new RegisterLink("L1_Status",    baseAddress_ + (0x10*addrSize), 1, 6,
                                "L1_GTXRdy",        Variable::Status, 0, 0x1,
                                "L1_DataValid",     Variable::Status, 1, 0x1, 
                                "L1_IlasActive",    Variable::Status, 2, 0x1,
                                "L1_nSync",         Variable::Status, 3, 0x1,                                                     
                                "L1_TxEnabled",     Variable::Status, 4, 0x1,
                                "L1_SysRefDetected",Variable::Status, 5, 0x1));
                                
   addRegisterLink(rl = new RegisterLink("L2_Status",     baseAddress_ + (0x11*addrSize), 1, 6,
                                "L2_GTXRdy",        Variable::Status, 0, 0x1,
                                "L2_DataValid",     Variable::Status, 1, 0x1, 
                                "L2_IlasActive",    Variable::Status, 2, 0x1,
                                "L2_nSync",         Variable::Status, 3, 0x1,                                                     
                                "L2_TxEnabled",     Variable::Status, 4, 0x1,
                                "L2_SysRefDetected",Variable::Status, 5, 0x1));
                                
    
   addRegisterLink(rl = new RegisterLink("L1_SignalSelect",    baseAddress_ + (0x20*addrSize), 1, 2,
                                "L1_data_out_mux",  Variable::Configuration, 0, 0xf,
                                "L1_test_out_mux",  Variable::Configuration, 4, 0xf));
                                
   rl->getVariable()->setDescription("Select between: b000 - Output zero, b001 - Parallel data from inside FPGA, b010 - Data from AXI stream, b011 - Test data ");                            
    
   addRegisterLink(rl = new RegisterLink("L2_SignalSelect",    baseAddress_ + (0x21*addrSize), 1, 2,
                                "L2_data_out_mux",  Variable::Configuration, 0, 0xf,
                                "L2_test_out_mux",  Variable::Configuration, 4, 0xf));
                                
   rl->getVariable()->setDescription("Select between: b000 - Output zero, b001 - Parallel data from inside FPGA, b010 - Data from AXI stream, b011 - Test data ");     
                                
   // Variables
   getVariable("Enabled")->setHidden(true);

   //Commands


}

// Deconstructor
JesdTx::~JesdTx ( ) { }

// Process Commands

File: python/surf/_JesdTx.py
#!/usr/bin/env python
#-----------------------------------------------------------------------------
# Title      : PyRogue JESD TX Module
#-----------------------------------------------------------------------------
# File       : JesdTx.py
# Created    : 2017-04-04
#-----------------------------------------------------------------------------
# Description:
# PyRogue JESD TX Module
#-----------------------------------------------------------------------------
# This file is part of the rogue software platform. It is subject to
# the license terms in the LICENSE.txt file found in the top-level directory
# of this distribution and at:
#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
# No part of the rogue software platform, including this file, may be
# copied, modified, propagated, or distributed except according to the terms
# contained in the LICENSE.txt file.
#-----------------------------------------------------------------------------

import pyrogue as pr

class JesdTx(pr.Device):
    def __init__(self, name="JesdTx", description="JESD TX Module", memBase=None, offset=0x0, hidden=False, numTxLanes=2):
        super(self.__class__, self).__init__(name, description, memBase, offset, hidden)

        digits = len(str(abs(numTxLanes-1)))

        ##############################
        # Variables
        ##############################

        self.add(pr.Variable(   name         = "Enable",
                                description  = "Enable mask. Example: 0x3 Enable ln0 and ln1.",
                                offset       =  0x00,
                                bitSize      =  numTxLanes,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "SubClass",
                                description  = "Jesd204b SubClass. 0 - For designs without sysref (no fixed latency). 1 - Fixed latency.",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "ReplaceEnable",
                                description  = "ReplaceEnable. Replace the control characters with data. (Should be 1 use 0 only for debug).",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x01,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "ResetGTs",
                                description  = "ResetGTs. Request reset of the GT modules.",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x02,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "ClearErrors",
                                description  = "Clear Jesd Errors and reset the status counters.",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x03,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "InvertSync",
                                description  = "InvertSync. Invert sync input (the AMC card schematics should be checkes if inverted). ",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x04,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "TestSigEnable",
                                description  = "Invert Sync. Sync output has to be inverted in some systems depending on signal polarities on the PCB.",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x05,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "ScrambleEnable",
                                description  = "ScrambleEnable. Enable data scrambling (More info in Jesd204b standard). ",
                                offset       =  0x10,
                                bitSize      =  1,
                                bitOffset    =  0x06,
                                base         = "enum",
                                mode         = "RW",
                                enum         = {
                                                  0 : "Disabled",
                                                  1 : "Enabled",
                                               },
                            ))

        self.add(pr.Variable(   name         = "RampStep",
                                description  = "Ramp increment step and a period of the wave in c-c",
                                offset       =  0x14,
                                bitSize      =  16,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "SquarePeriod",
                                description  = "Ramp increment step and a period of the wave in c-c",
                                offset       =  0x16,
                                bitSize      =  16,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "LowAmplitudeVal",
                                description  = "Low value of the square waveform amplitude",
                                offset       =  0x18,
                                bitSize      =  32,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "HighAmplitudeVal",
                                description  = "High value of the square waveform amplitude",
                                offset       =  0x1C,
                                bitSize      =  32,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        self.add(pr.Variable(   name         = "InvertDacData",
                                description  = "Mask Enable the DAC data inversion. 1-Inverted, 0-normal.",
                                offset       =  0x20,
                                bitSize      =  numTxLanes,
                                bitOffset    =  0x00,
                                base         = "hex",
                                mode         = "RW",
                            ))

        for i in range(numTxLanes):
            self.add(pr.Variable(   name         = "GTReady_%.*i" % (digits, i),
                                    description  = "GT Ready. Jesd clock ok PLLs are locked and GT is ready to receive data. Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x00,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "DataValid_%.*i" % (digits, i),
                                    description  = "Jesd Data Valid. Goes high after the code synchronisation and ILAS sequence is complete (More info in Jesd204b standard). Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x01,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "IlasActive_%.*i" % (digits, i),
                                    description  = "ILA sequence Active. Only 1 for 4 multiframe clock cycles then it drops (More info in Jesd204b standard). Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x02,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "nSync_%.*i" % (digits, i),
                                    description  = "nSync. 0 - Not synchronised. 1 - Indicades that code group synchronisation has been completed. Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x03,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "TxEnabled_%.*i" % (digits, i),
                                    description  = "Tx Lane Enabled. Indicates if the lane had been enabled in configuration. Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x04,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "SysRefDetected_%.*i" % (digits, i),
                                    description  = "System Reference input has been Detected. Lane %.*i" % (digits, i),
                                    offset       =  0x40 + (0x04*i),
                                    bitSize      =  1,
                                    bitOffset    =  0x05,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

            self.add(pr.Variable(   name         = "dataOutMux_%.*i" % (digits, i),
                                    description  = "data_out_mux: Select between: b000 - Output zero, b001 - Parallel data from inside FPGA, b010 - Data from AXI stream (not used), b011 - Test data. Lane %.*i" % (digits, i),
                                    offset       =  0x80 + (0x04*i),
                                    bitSize      =  4,
                                    bitOffset    =  0x00,
                                    base         = "enum",
                                    mode         = "RW",
                                    enum         = {
                                                      0 : "OutputZero",
                                                      1 : "UserData",
                                                      2 : "AxiStream",
                                                      3 : "TestData",
                                                   },
                                ))

            self.add(pr.Variable(   name         = "testOutMux_%.*i" % (digits, i),
                                    description  = "test_out_mux[1:0]: Select between: b000 - Saw signal increment, b001 - Saw signal decrement, b010 - Square wave,  b011 - Output zero. Lane %.*i" % (digits, i),
                                    offset       =  0x80 + (0x04*i),
                                    bitSize      =  4,
                                    bitOffset    =  0x04,
                                    base         = "enum",
                                    mode         = "RW",
                                    enum         = {
                                                      0 : "SawIncrement",
                                                      1 : "SawDecrement",
                                                      2 : "SquareWave",
                                                      3 : "OutputZero",
                                                   },
                                ))

            self.add(pr.Variable(   name         = "StatusValidCnt_%.*i" % (digits, i),
                                    description  = "StatusValidCnt[31:0]. Shows stability of JESD lanes. Counts number of JESD re-syncronisations. Lane %.*i" % (digits, i),
                                    offset       =  0x100 + (0x04*i),
                                    bitSize      =  32,
                                    bitOffset    =  0x00,
                                    base         = "hex",
                                    mode         = "RO",
                                ))

        ##############################
        # Commands
        ##############################

        self.add(pr.Command (   name         = "ClearTxStatus",
                                description  = "Clear the status valid counter of TX lanes.",
                                function     = """\
                                               self.ClearErrors.set(1)
                                               self.ClearErrors.set(0)
                                               """
                            ))

        self.add(pr.Command (   name         = "ResetTxGTs",
                                description  = "Toggle the reset of all TX MGTs",
                                function     = """\
                                               self.ResetGTs.set(1)
                                               self.ResetGTs.set(0)
                                               """
                            ))

